### PR TITLE
Use queue to prevent 429 returns

### PR DESCRIPTION
**Pull request by pirquessa**
In order to do faster checks (#12) and prevent 429 returns, let's use a job queue

### CODE REVIEW

File: background_scripts/index.js [hunk 1]
```diff
@@ -1,5 +1,9 @@
 // Scan périodique des RDV
 (async function () {
+  const TIME_BETWEEN_JOBS = 20;
```
Comment (pirquessa) [inline background_scripts/index.js:3]: The timer to use, we might want to make it dependant of the queue length

File: background_scripts/index.js [hunk 2]
```diff
@@ -10,12 +14,30 @@
...
-    const $iframe = document.createElement("iframe");
-    // On charge l'URL dans une iframe
-    // Ici on laisse la main au content script qui va vérifier si un RDV est disponible
-    $iframe.src = url;
-    $iframe.id = url;
-    document.body.appendChild($iframe);
+    const iframe = document.createElement("iframe");
+    document.body.appendChild(iframe);
+    
+    return iframe;
```
Comment (pirquessa) [inline background_scripts/index.js:20]: We create empty iframe, a job will "run" them

File: background_scripts/index.js [hunk 3]
```diff
@@ -10,12 +14,30 @@
...
+
+    const job = jobs.shift();
+    if (job) {
+      console.info('Start job ' + job);
+      
+      const iframe = iframes[job];
+      // On charge l'URL dans une iframe
+      // Ici on laisse la main au content script qui va vérifier si un RDV est disponible
+      iframe.src = job;
+    }
```
Comment (pirquessa) [inline background_scripts/index.js:40]: If processing is not stopped, find next job and execute it

File: background_scripts/index.js [hunk 4]
```diff
@@ -45,6 +67,13 @@
     console.info(data);
 
     switch (data.type) {
+      case "over":
+        // Close Iframe
+        iframes[data.url].src = "";
+        // Set job in the queue for next execution
+        jobs.push(data.url);
+        break;
```
Comment (pirquessa) [inline background_scripts/index.js:75]: When a job is over, clean it, add it to the queue again

File: background_scripts/index.js [hunk 5]
```diff
@@ -87,5 +116,12 @@
   await updateIconStatus(stopped);
 
   // On ajoute les iframes pour charger les centres à surveiller en arrière plan
-  Object.keys(locations).forEach((url) => createIframe(url));
+  Object.keys(locations).forEach((url) => {
+    iframes[url] = createIframe(url);
+    jobs.push(url);
+  });
```
Comment (pirquessa) [inline background_scripts/index.js:122]: keep track of iframes and init first jobs

File: background_scripts/index.js [hunk 6]
```diff
@@ -87,5 +116,12 @@
...
+
+  // Execute jobs every TIME_BETWEEN_JOBS sec
+  setInterval(executeNextJob, TIME_BETWEEN_JOBS * 1000);
+  executeNextJob();
```
Comment (pirquessa) [inline background_scripts/index.js:126]: Execute jobs and start one now !

File: content_scripts/doctolib/book.js [hunk 7]
```diff
@@ -322,10 +322,14 @@
...
-        },
+        }
       });
     } finally {
-      setTimeout(window.location.reload.bind(window.location), 3 * 60 * 1000);
+      browser.runtime.sendMessage({
+        type: "over",
+        url,
+        location: locations[url]
+      });
```
Comment (pirquessa) [inline content_scripts/doctolib/book.js:332]: Inform master that this job is over

**Comment by dunglas** [comment]
LGTM! Have you tested this patch for real to make sure it works?

**Comment by pirquessa** [comment]
> LGTM! Have you tested this patch for real to make sure it works?

Quickly, I don't master all your project ;-) And it's my first contribution to browser plugin.

On peut aussi discuter en FR, j'ai l'habitude de bosser en EN... c'est peut etre pas utile ici. En tout cas, mon dernier commit 4d44490 corrige un bug important: on check si les slots sont dispo avant qu'ils soient chargés dans la UI... Il faudrait peut etre prendre celui-ci au minimum.

**Pull request closed without merging**